### PR TITLE
Enable long aliases

### PR DESCRIPTION
**Pull request by Trenly**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Are you working against an Issue?

This change allows for argument aliases to be more than 1 character

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2107)

### CODE REVIEW

**Comment by Trenly** [comment]
> Supporting multi-character aliases removes, or at least complicates greatly, this feature:

I did not realize that was a feature. You are correct that I did not consider it here.

> What is the goal in having longer aliases?

The goal here is to allow overall extended functionality. I had simply used verbose as an example. There are use cases where a single letter alias doesn’t make sense, or would conflict with an alias of an already existing argument. This would allow for a wider range of aliases to be available.

A secondary effect here is to make it easier to use in general. Looking from the perspective of a new user, and just as an example, I wouldn’t know that `--verbose-logs` was a thing, but I would expect (based on powershell and other conventions) that `-verbose` would be a standard behavior.

**Comment by Trenly** [comment]
> But I wouldn't be opposed to an arcane alias for it seeing as it has none today. Something like `-+` makes some sense, and is not likely that anyone else would want to use it.

I've updated the PR so that any aliases longer than 1 character require the `-+` syntax

The other thought I had was that perhaps we could still use `--` for long aliases, since at that point it would just be a short argument rather than a long alias;

**Comment by JohnMcPMS** [comment]
> I've updated the PR so that any aliases longer than 1 character require the `-+` syntax

I meant that the alias for the verbose logs option would be the character `+`, not that a new syntax should be created.

> The other thought I had was that perhaps we could still use `--` for long aliases, since at that point it would just be a short argument rather than a long alias;

I am not opposed to alternate names for arguments.  You don't even have to write new code for that necessarily (although it is probably the better solution).

File: src/AppInstallerCLICore/Argument.h [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 #define APPINSTALLER_CLI_ARGUMENT_IDENTIFIER_CHAR       '-'
 #define APPINSTALLER_CLI_ARGUMENT_IDENTIFIER_STRING     "-"
+#define APPINSTALLER_CLI_LONG_ALIAS_IDENTIFIER_CHAR     '+'
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Argument.h:17]: Not a fan of adding a new syntax here; would prefer the alternate name strategy.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]
Sorry about the delay in reviewing.  I do think that the easiest way forward is allowing for alternate names for arguments while leaving the alias character stuff alone.

**Comment by Trenly** [comment]
> I meant that the alias for the verbose logs option would be the character `+`, not that a new syntax should be created.
> 
Ahhh, I misunderstood. Removed.

> I am not opposed to alternate names for arguments. You don't even have to write new code for that necessarily (although it is probably the better solution).
>
Made this the functionality. Aliases longer than 1 character now are functionally equivalent to alternate names

File: src/AppInstallerCLICore/Argument.h [hunk 2]
```diff
@@ -47,24 +47,24 @@ namespace AppInstaller::CLI
         };
 
         // Defines an argument with no alias.
-        constexpr static char NoAlias = '\0';
+        constexpr static char const* NoAlias = "\0";
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Argument.h:50]:
```suggestion
        constexpr static std::string_view NoAlias;
```
Can just be empty string.

File: src/AppInstallerCLICore/Argument.h [hunk 3]
```diff
@@ -47,24 +47,24 @@ namespace AppInstaller::CLI
         };
 
         // Defines an argument with no alias.
-        constexpr static char NoAlias = '\0';
+        constexpr static char const* NoAlias = "\0";
 
-        Argument(std::string_view name, char alias, Execution::Args::Type execArgType, Resource::StringId desc) :
+        Argument(std::string_view name, char const* alias, Execution::Args::Type execArgType, Resource::StringId desc) :
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Argument.h:52]:
```suggestion
        Argument(std::string_view name, std::string_view alias, Execution::Args::Type execArgType, Resource::StringId desc) :
```
Although mildly more dangerous than `char*`, these should all be `std::string_view`, as should the storage and return value from the getter.

Also consider keeping `char Alias() const` and adding `std::string_view AlternateName() const`.  Then `Alias` returns `NoAlias` if the stored string view is not exactly 1 character, or the single character if it is 1.  `AlternateName` returns a non-empty view only if the stored view is > 1 character.  That way, the consuming code doesn't need to understand how we decide between the two options, it just needs to know how to tell if either of the values is valid or not.

File: src/AppInstallerCLICore/Command.cpp [hunk 4]
```diff
@@ -226,7 +230,14 @@ namespace AppInstaller::CLI
                 std::ostringstream strstr;
                 if (arg.Alias() != Argument::NoAlias)
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Command.cpp:231]: Mildly dangerous check here if you don't change to `std::string_view`.  Comparing pointers only works for the literals.

File: src/AppInstallerCLICore/Command.cpp [hunk 5]
```diff
@@ -492,7 +503,7 @@ namespace AppInstaller::CLI
         else if (currArg[1] != APPINSTALLER_CLI_ARGUMENT_IDENTIFIER_CHAR)
         {
             // Parse the single character alias argument
-            char currChar = currArg[1];
+            std::string currChar(1, currArg.at(1));
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Command.cpp:506]: Is there any value to this change now?

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCLICore/Argument.h (hunk 3 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Argument.h]: I've added the alternate name as a member variable.

For the sake of not modifying every single Argument command again, I've created additional overloads for the constructor to handle these. If this is not the appropriate way to handle this, let me know so I can fix it accordingly

Thanks for your guidance and patience 😄

File: src/AppInstallerCLICore/Argument.cpp [hunk 6]
```diff
@@ -96,7 +96,7 @@ namespace AppInstaller::CLI
         case Args::Type::Wait:
             return Argument{ "wait", NoAlias, Args::Type::Wait, Resource::String::WaitArgumentDescription, ArgumentType::Flag, false };
         case Args::Type::ProductCode:
-            return Argument{ "product-code", NoAlias, Args::Type::ProductCode, Resource::String::ProductCodeArgumentDescription, ArgumentType::Standard, false };
+            return Argument{ "product-code", NoAlias, "pid", Args::Type::ProductCode, Resource::String::ProductCodeArgumentDescription, ArgumentType::Standard, false };
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Argument.cpp:99]: `pid` usually means "Process ID".  I don't see how it relates to product code.

File: src/AppInstallerCLICore/Argument.h [hunk 7]
```diff
@@ -49,23 +49,44 @@ namespace AppInstaller::CLI
         // Defines an argument with no alias.
         constexpr static char NoAlias = '\0';
 
+        // Defines an argument with no alternate name
+        constexpr static std::string_view NoAlternateName = "";
+
         Argument(std::string_view name, char alias, Execution::Args::Type execArgType, Resource::StringId desc) :
-            m_name(name), m_alias(alias), m_execArgType(execArgType), m_desc(std::move(desc)) {}
+            m_name(name), m_alias(alias), m_alternateName(NoAlternateName), m_execArgType(execArgType), m_desc(std::move(desc)) {}
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Argument.h:56]: nit: The default constructed state of a `std::string_view` is already an empty string.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]
The test that ensures that there are no commands with arguments that collide should be updated to check against `AlternateName` as well.

File: src/AppInstallerCLICore/Argument.cpp (hunk 6 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Argument.cpp:99]: `pid` has multiple meanings. In the terms of a running service, it does typically mean "Process ID". I've seen it used in the application space to refer to "Product ID"

Is there a better alternate name to use here, or should I just remove this one?

Comment (JohnMcPMS) [reply src/AppInstallerCLICore/Argument.cpp:99]: I find it to be confusing, I would remove it as I don't foresee this being a commonly used argument in need of shortening.

File: src/AppInstallerCLITests/Command.cpp [hunk 8]
```diff
@@ -69,6 +81,7 @@ void EnsureCommandConsistency(const Command& command)
     auto args = command.GetArguments();
     Argument::GetCommon(args);
     EnsureStringsAreLowercaseAndNoCollisions(command.FullName() + " argument names", args, GetArgumentName);
+    EnsureStringsAreLowercaseAndNoCollisions(command.FullName() + " argument alternate name", args, GetArgumentAlternateName);
```
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerCLITests/Command.cpp:84]: Each one of the current calls to `EnsureStringsAreLowercaseAndNoCollisions` will detect collisions within their respective space, `Name` and `AlternateName`.  They do that by simply creating a set in memory and looking for duplicates as they go.  I'm suggesting that we can pass the set in and thus preserve it across the two calls:
```
std::unordered_set<std::string> allNames;
EnsureStringsAreLowercaseAndNoCollisions(command.FullName() + " argument names", args, GetArgumentName, allNames);
EnsureStringsAreLowercaseAndNoCollisions(command.FullName() + " argument alternate name", args, GetArgumentAlternateName, allNames);
```

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**